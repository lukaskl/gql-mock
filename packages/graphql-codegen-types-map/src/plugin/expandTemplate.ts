import * as changeCase from 'change-case'
import { parse } from 'esprima'
import { OperationTypeNode } from 'graphql'
import evaluate from 'static-eval'

export interface TemplateVariables {
  operationName: string
  operationKind: OperationTypeNode
}

const casingOperations = (Object.keys(changeCase) as (keyof typeof changeCase)[])
  .map(x => ({ [x]: changeCase[x] }))
  .reduce((l, r) => ({ ...l, ...r }), {})

export const expandTemplate = (template: string, variables: TemplateVariables) => {
  const ast = (parse('`' + template.replace(/{/g, '${') + '`').body[0] as any).expression

  const result: string = evaluate(ast, {
    ...casingOperations,
    ...variables,
  })
  if (result.includes('[object Object]')) {
    const errorMessage =
      `Invalid variable or function name used in "operationsMap" template. ` +
      `Given template: "${template}". ` +
      `Allowed variables: "${Object.keys(variables).join(', ')}". ` +
      `Allowed functions: "${Object.keys(casingOperations).join(', ')}". ` +
      `e.g. template: "{pascalCase(operationName)}{pascalCase(operationKind)}Variables"`
    console.error(`ERROR: ${errorMessage}`)
    throw new Error(errorMessage)
  }
  return result
}

export const validateTemplates = (...templates: string[]) => {
  const validationVariables: TemplateVariables = {
    operationKind: 'query',
    operationName: 'validation',
  }
  for (const template of templates) {
    expandTemplate(template, validationVariables)
  }
}
