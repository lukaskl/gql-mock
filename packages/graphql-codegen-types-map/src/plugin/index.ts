export { plugin } from './plugin'
export { OperationsParser } from './parser'
