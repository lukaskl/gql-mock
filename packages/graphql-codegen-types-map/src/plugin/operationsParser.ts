import {
  ASTNode,
  DocumentNode,
  VariableDefinitionNode,
  visit,
  OperationTypeNode,
  print,
  TypeInfo,
  visitWithTypeInfo,
  BREAK,
} from 'graphql'

export interface OperationsMapEntry {
  operationName: string
  hasVariables: boolean
  operationKind: OperationTypeNode
}

export const getVariablesAst = (node: ASTNode) => {
  let variable: VariableDefinitionNode | undefined = undefined
  visit(node, {
    VariableDefinition: node => {
      variable = node
      return BREAK
    },
  })

  return variable
}

export const getOperationsMap = (typeInfo: TypeInfo, allAst: DocumentNode) => {
  const operations: OperationsMapEntry[] = []
  visit(
    allAst,
    visitWithTypeInfo(typeInfo, {
      leave: {
        OperationDefinition: node => {
          const name = node.name?.value
          if (!name) {
            throw new Error(`Anonymous operations are not allowed: ${print(node)}`)
          }
          const operationKind = node?.operation
          const variables = getVariablesAst(node)
          operations.push({
            operationName: name,
            hasVariables: !!variables,
            operationKind,
          })
        },
      },
    })
  )
  return operations
}
