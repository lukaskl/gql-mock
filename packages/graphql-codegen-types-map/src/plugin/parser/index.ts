export { OperationsParser } from './OperationsParser'
export { FragmentUsage, TypeUsages, getTypeUsages, UsagePathNode } from './getTypeUsages'
