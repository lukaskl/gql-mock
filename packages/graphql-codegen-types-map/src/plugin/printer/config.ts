import {
  validateFieldArgsTemplate,
  validateOperationsTemplate,
  validateTypeAccessorTemplate,
} from './expandTemplate'
import merge from 'lodash.merge'
import { pickKeys, KeysObj } from '~/utils'

export interface OperationsTemplatesConfig {
  operationTypeTemplate: string
  operationDocumentTemplate: string
  variablesTypeTemplate: string
  operationKindTemplate: string
  typeUsagesTemplate: string
}
export interface FieldArgsTemplateConfig {
  fieldArgsTypeTemplate: string
}

export interface TypeAccessorTemplateConfig {
  typeAccessorTypeTemplate: string
}

export type TemplatesConfig = OperationsTemplatesConfig & FieldArgsTemplateConfig & TypeAccessorTemplateConfig

export interface TypesImportConfig {
  importTypesFrom: string | undefined
  importedTypesAlias: string
}

export interface TypeUsagesConfig extends Pick<OperationsTemplatesConfig, 'typeUsagesTemplate'> {
  withTypeUsages: boolean
}

export interface AllConfigOptions extends TemplatesConfig, TypesImportConfig, TypeUsagesConfig {
  withDocumentsMap: boolean
}

export const defaultConfig: AllConfigOptions = {
  operationTypeTemplate: '{OperationName}{OperationKind}',
  variablesTypeTemplate: '{OperationName}{OperationKind}Variables',
  operationDocumentTemplate: '{OperationName}{OperationKind === "Fragment" ? "FragmentDoc" : "Document"}',
  operationKindTemplate: '{operationKind}',
  typeAccessorTypeTemplate: '{typeKind === "scalar" ? `Scalars["{typeName}"]` : TypeName }',
  importedTypesAlias: 'Types',
  importTypesFrom: undefined,
  typeUsagesTemplate: 'TypeUsagesFor{OperationName}{OperationKind}',
  fieldArgsTypeTemplate: '{ParentName}{FieldName}Args',
  withTypeUsages: true,
  withDocumentsMap: true,
}

const operationsTemplatesConfigKeysObs: KeysObj<keyof OperationsTemplatesConfig> = {
  operationKindTemplate: null,
  operationTypeTemplate: null,
  variablesTypeTemplate: null,
  typeUsagesTemplate: null,
  operationDocumentTemplate: null,
}
const fieldArgsTemplatesConfigKeysObs: KeysObj<keyof FieldArgsTemplateConfig> = {
  fieldArgsTypeTemplate: null,
}
const typeAccessorTemplatesConfigKeysObs: KeysObj<keyof TypeAccessorTemplateConfig> = {
  typeAccessorTypeTemplate: null,
}
const typesImportConfigKeysObs: KeysObj<keyof TypesImportConfig> = {
  importTypesFrom: null,
  importedTypesAlias: null,
}
const typeUsagesConfigKeysObs: KeysObj<keyof TypeUsagesConfig> = {
  withTypeUsages: null,
  typeUsagesTemplate: null,
}

export class PrinterConfig {
  private readonly config: AllConfigOptions

  constructor(rawConfig: Partial<AllConfigOptions> = {}) {
    this.config = merge(defaultConfig, rawConfig)
  }

  get allOptions() {
    return this.config
  }

  get operationTemplates() {
    return this.pickConfigEntries(operationsTemplatesConfigKeysObs)
  }

  get fieldArgsTemplates() {
    return this.pickConfigEntries(fieldArgsTemplatesConfigKeysObs)
  }

  get typeAccessorTemplates() {
    return this.pickConfigEntries(typeAccessorTemplatesConfigKeysObs)
  }

  get importTypes() {
    const config = this.pickConfigEntries(typesImportConfigKeysObs)
    const importRef = config.importTypesFrom ? `${config.importedTypesAlias}.` : ''
    return { ...config, importRef }
  }

  get typeUsages() {
    return this.pickConfigEntries(typeUsagesConfigKeysObs)
  }

  validateConfig() {
    validateOperationsTemplate(...Object.values(this.operationTemplates))
    validateFieldArgsTemplate(...Object.values(this.fieldArgsTemplates))
    validateTypeAccessorTemplate(...Object.values(this.typeAccessorTemplates))
  }

  private pickConfigEntries = <T extends keyof AllConfigOptions>(pick: KeysObj<T>) =>
    pickKeys<T, AllConfigOptions>(this.config, pick)
}
