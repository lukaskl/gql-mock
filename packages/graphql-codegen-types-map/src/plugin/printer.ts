import { expandTemplate } from './expandTemplate'
import { OperationsMapEntry } from './operationsParser'
import { TemplatesConfig } from './plugin'

export interface OperationsMapPrinterConfig {
  operationsMap?: {
    operationKindTemplate?: string
    operationTypeTemplate?: string
    variablesTypeTemplate?: string
  }
}

export class OperationsPrinter {
  constructor(protected readonly operations: OperationsMapEntry[], readonly templates: TemplatesConfig) {}

  get operationsMapFields() {
    const { operationKindTemplate, operationTypeTemplate, variablesTypeTemplate } = this.templates

    return this.operations
      .map(x => {
        const opKind = expandTemplate(operationKindTemplate, x)
        const opType = expandTemplate(operationTypeTemplate, x)
        const varType = expandTemplate(variablesTypeTemplate, x)

        return `'${x.operationName}': { operationType: ${opType}, variablesType: ${varType}, kind: '${opKind}' }`
      })
      .join('\n  ')
  }

  get operationsMapInterface() {
    return `export interface OperationsMap {\n  ${this.operationsMapFields}\n}`
  }
}
