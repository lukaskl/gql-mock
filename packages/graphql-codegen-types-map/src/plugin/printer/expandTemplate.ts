import * as changeCase from 'change-case'
import { parse } from 'esprima'
import { OperationTypeNode } from 'graphql'
import evaluate from 'static-eval'
import { ERRORS, fromEntries } from '~/utils'
import { GraphQLSchemaOutputTypes } from '../parser'

const casingOperations = (Object.keys(changeCase) as (keyof typeof changeCase)[])
  .map(x => ({ [x]: changeCase[x] }))
  .reduce((l, r) => ({ ...l, ...r }), {})

export type OperationKind = OperationTypeNode | 'fragment'
export interface OperationTemplateVariables {
  operationName: string
  operationKind: OperationKind
}

export interface FieldArgsTemplateVariables {
  parentName: string
  fieldName: string
}

export interface TypeAccessorTemplateVariables {
  typeName: string
  typeKind: GraphQLSchemaOutputTypes
}

export type TemplateVariables =
  | FieldArgsTemplateVariables
  | OperationTemplateVariables
  | TypeAccessorTemplateVariables

export const expandTemplate = <T extends TemplateVariables>(template: string, variables: T): string => {
  const ast = (parse('`' + template.replace(/{/g, '${') + '`').body[0] as any).expression

  const allVariables = {
    ...fromEntries(
      Object.entries(variables).map(([name, value]) => [
        changeCase.pascalCase(name),
        changeCase.pascalCase(value),
      ])
    ),
    ...variables,
  }

  const result: string = evaluate(ast, {
    ...casingOperations,
    ...allVariables,
  })
  if (result.includes('[object Object]')) {
    return ERRORS.invalidTemplate(template, allVariables, casingOperations)
  }
  return result
}

const validateTemplate = (templates: string[], dummyVariables: TemplateVariables) => {
  for (const template of templates) {
    expandTemplate(template, dummyVariables)
  }
}

export const validateOperationsTemplate = (...templates: string[]) =>
  validateTemplate(templates, {
    operationKind: 'query',
    operationName: 'validation',
  })

export const validateFieldArgsTemplate = (...templates: string[]) =>
  validateTemplate(templates, {
    fieldName: 'field-name',
    parentName: 'parent-name',
  })

export const validateTypeAccessorTemplate = (...templates: string[]) =>
  validateTemplate(templates, {
    typeKind: 'interface',
    typeName: 'fake-name',
  })
