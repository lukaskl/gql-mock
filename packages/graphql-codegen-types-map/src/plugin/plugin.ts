import { PluginFunction } from '@graphql-codegen/plugin-helpers'
import { concatAST, DocumentNode, TypeInfo } from 'graphql'

import { validateTemplates } from './expandTemplate'
import { getOperationsMap } from './operationsParser'
import { OperationsMapPrinterConfig, OperationsPrinter } from './printer'

const ensureConfigDefaults = (config: OperationsMapPrinterConfig) => {
  const {
    operationsMap: {
      operationTypeTemplate = '{pascalCase(operationName)}{pascalCase(operationKind)}',
      variablesTypeTemplate = '{pascalCase(operationName)}{pascalCase(operationKind)}Variables',
      operationKindTemplate = '{operationKind}',
    } = {},
  } = config

  validateTemplates(operationTypeTemplate, variablesTypeTemplate, operationKindTemplate)
  return { operationTypeTemplate, variablesTypeTemplate, operationKindTemplate }
}

export type TemplatesConfig = ReturnType<typeof ensureConfigDefaults>

export const plugin: PluginFunction<OperationsMapPrinterConfig> = (schema, rawDocuments, config) => {
  const templates = ensureConfigDefaults(config)
  const typeInfo = new TypeInfo(schema)

  const documents = rawDocuments
  const allAst = concatAST(
    documents
      .map(
        v =>
          v.document ||
          // The next line is to make it work with older versions (<= v1.6.1) of @graphql-codegen
          (v as any).content
      )
      .filter(x => !!x) as DocumentNode[]
  )

  const operations = getOperationsMap(typeInfo, allAst)
  const operationsMapPrinter = new OperationsPrinter(operations, templates)

  return operationsMapPrinter.operationsMapInterface
}
