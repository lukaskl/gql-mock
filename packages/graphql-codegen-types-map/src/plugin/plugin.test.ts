import { parse, DocumentNode } from 'graphql'
import { schema } from '~/test-support'
import { plugin } from './plugin'
import { OperationsMapPrinterConfig } from './printer'

const runPlugin = async (ast: DocumentNode | DocumentNode[], config: OperationsMapPrinterConfig = {}) => {
  const asts = Array.isArray(ast) ? ast : [ast]
  return await plugin(
    schema,
    asts.map(x => ({ location: 'test-file.ts', document: x })),
    config,
    {
      outputFile: '',
    }
  )
}

describe('TypeScript Operations Map Plugin', () => {
  it('Anonymous operation throws an error', async () => {
    const ast = parse(/* GraphQL */ `
      query {
        notifications {
          id
        }
      }
    `)

    await expect(runPlugin(ast)).rejects.toThrow()
  })

  it('Test 1', async () => {
    const ast = parse(/* GraphQL */ `
      query q1 {
        search {
          ... on Movie {
            __typename
            id
            title
          }
          ... on Person {
            __typename
            id
            name
          }
        }
      }

      query q2 {
        search {
          __typename
          ... on Movie {
            id
            title
          }
          ... on Person {
            id
            name
          }
        }
      }
    `)
    const result = await runPlugin(ast)

    expect(result).toMatchSnapshot()
  })

  it('Test 2', async () => {
    const ast = parse(/* GraphQL */ `
      query search($trms: String!) {
        search(term: $trms) {
          id
        }
      }
    `)

    const result = await runPlugin(ast)

    expect(result).toMatchSnapshot()
  })

  it('Test 3', async () => {
    const ast1 = parse(/* GraphQL */ `
      query search1($trms: String!) {
        search(term: $trms) {
          id
        }
      }
    `)
    const ast2 = parse(/* GraphQL */ `
      query search2($trms: String!) {
        search(term: $trms) {
          id
        }
      }
    `)

    const result = await runPlugin([ast1, ast2])

    expect(result).toMatchSnapshot()
  })
})
