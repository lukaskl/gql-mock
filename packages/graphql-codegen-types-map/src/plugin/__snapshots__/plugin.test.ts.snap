// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript Operations Map Plugin Test 1 1`] = `
"export interface OperationsMap {
  'q1': { operationType: Q1Query, variablesType: Q1QueryVariables, kind: 'query' }
  'q2': { operationType: Q2Query, variablesType: Q2QueryVariables, kind: 'query' }
}"
`;

exports[`TypeScript Operations Map Plugin Test 2 1`] = `
"

export interface FieldArgsUsagesMap {
  'Query': {
    search: QuerySearchArgs
  },
  'Mutation': {
    login: MutationLoginArgs
  }
}

export type TypeUsagesForSearchQuery = { SearchResult: SearchQuery['search'][0] } &
  { Query: SearchQuery }

export interface OperationsMap {
  'search': { operationType: SearchQuery, variablesType: SearchQueryVariables, typeUsages: TypeUsagesForSearchQuery, kind: 'query' }
}

export interface TypesMap {
  operations: OperationsMap
  fieldArgsUsages: FieldArgsUsagesMap
}

export const documentsMap = {
  'search': { document: SearchDocument, kind: 'query' as const }
}"
`;

exports[`TypeScript Operations Map Plugin Test 3 1`] = `
"

export interface FieldArgsUsagesMap {
  'Query': {
    search: QuerySearchArgs
  },
  'Mutation': {
    login: MutationLoginArgs
  }
}

export type TypeUsagesForSearch1Query = { SearchResult: Search1Query['search'][0] } &
  { Query: Search1Query }

  export type TypeUsagesForSearch2Query = { SearchResult: Search2Query['search'][0] } &
  { Query: Search2Query }

export interface OperationsMap {
  'search1': { operationType: Search1Query, variablesType: Search1QueryVariables, typeUsages: TypeUsagesForSearch1Query, kind: 'query' }
  'search2': { operationType: Search2Query, variablesType: Search2QueryVariables, typeUsages: TypeUsagesForSearch2Query, kind: 'query' }
}

export interface TypesMap {
  operations: OperationsMap
  fieldArgsUsages: FieldArgsUsagesMap
}

export const documentsMap = {
  'search1': { document: Search1Document, kind: 'query' as const },
  'search2': { document: Search2Document, kind: 'query' as const }
}"
`;
